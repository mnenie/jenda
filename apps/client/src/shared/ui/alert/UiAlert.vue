<script setup lang="ts">
import { computed, type HTMLAttributes, ref } from 'vue'
import { alertVariants, type AlertVariants } from '.'
import { cn } from '@/shared/libs/shadcn/utils'

interface AlertProps {
  variant?: AlertVariants['variant']
  closable?: boolean
  content?: string
  class?: HTMLAttributes['class']
}

const { variant = 'default', closable, content, class: alertClass } = defineProps<AlertProps>()

const emits = defineEmits(['close'])

const isAlertVisible = ref(true)

const valueOfRightPadding = computed(() => (closable ? '32px' : '10px'))
</script>

<template>
  <div
    v-if="isAlertVisible"
    :class="cn(
      'alert',
      alertVariants({ variant }),
      alertClass,
    )"
    :style="{ 'padding-right': valueOfRightPadding }"
  >
    <slot v-if="$slots.default || content">
      <span class="text-default">{{ content }}</span>
    </slot>
    <div
      v-if="closable"
      i-lucide-x
      class="alert-close"
      @click="isAlertVisible = false, emits('close')"
    />
  </div>
</template>
