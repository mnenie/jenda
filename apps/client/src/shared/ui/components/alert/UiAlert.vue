<script setup lang="ts">
import { computed, type HTMLAttributes } from 'vue'
import { createReusableTemplate } from '@vueuse/core'
import { alertVariants, type AlertVariants } from '.'
import { cn } from '@/shared/libs/shadcn/utils'

interface AlertProps {
  variant?: AlertVariants['variant']
  closable?: boolean
  content?: string
  rp?: string
  withAnimation?: boolean
  class?: HTMLAttributes['class']
}

const { variant = 'default', class: alertClass, ...props } = defineProps<AlertProps>()

const emits = defineEmits(['close'])

const isAlertVisible = defineModel<boolean>('open', {
  default: true,
})

const valueOfRightPadding = computed(() => (props.rp ? '' : props.closable ? '32px' : '10px'))

function closeAlert() {
  isAlertVisible.value = false
  emits('close')
}

const [DefineTemplate, ReuseTemplate] = createReusableTemplate()
</script>

<template>
  <DefineTemplate>
    <div
      v-if="isAlertVisible"
      :class="cn(
        'alert',
        alertVariants({ variant }),
        alertClass,
      )"
      :style="{ 'padding-right': valueOfRightPadding }"
    >
      <slot v-if="$slots.default || content">
        <span class="text-default">{{ content }}</span>
      </slot>
      <div
        v-if="closable"
        i-lucide-x
        class="alert-close"
        @click="closeAlert"
      />
    </div>
  </DefineTemplate>
  <Transition v-if="withAnimation" name="slide-fade">
    <ReuseTemplate />
  </Transition>
  <ReuseTemplate v-else />
</template>

<style lang="css" scoped>
.slide-fade-enter-active,
.slide-fade-leave-active {
  transition:
    opacity 0.15s ease-out,
    transform 0.15s ease-out;
  will-change: transform, opacity;
}

.slide-fade-enter-from {
  transform: translateY(20px) translateX(-50%);
  opacity: 0;
}

.slide-fade-leave-to {
  transform: translateY(20px) translateX(-50%);
  opacity: 0;
}
</style>
