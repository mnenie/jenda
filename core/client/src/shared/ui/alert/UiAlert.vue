<script setup lang="ts">
import type { AlertVariant } from './types'
import { useDark } from '@vueuse/core'
import { computed } from 'vue'
import { variants } from '.'

interface AlertProps {
  variant?: AlertVariant
  closable?: boolean
  content?: string
}

const { variant = 'default', closable, content } = defineProps<AlertProps>()

const emits = defineEmits(['close'])

const isDark = useDark()

const _styles = computed(() => {
  return variants[variant]
})

const alertBgColor = computed(() =>
  isDark.value ? _styles.value.dark.background : _styles.value.light.background,
)
const alertBorder = computed(() => (isDark.value ? _styles.value.dark.border : _styles.value.light.border))

const valueOfRightPadding = computed(() => (closable ? '32px' : '10px'))
</script>

<template>
  <div :class="$style.alert" :style="{ background: alertBgColor, border: alertBorder }">
    <slot v-if="$slots.default || content">
      <span class="text-sm">{{ content }}</span>
    </slot>
    <div v-if="closable" i-lucide-x text-sm :class="$style.x_mark" @click="emits('close')" />
  </div>
</template>

<style module lang="scss">
.alert {
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 10px;
  padding: 2px v-bind('valueOfRightPadding') 2px 10px;
  border-radius: 8px;
  margin-bottom: 20px;
  min-height: 22px;

  .x_mark {
    position: absolute;
    top: 50%;
    transform: translateY(-50%);
    right: 10px;
    color: var(--zinc-600);
    cursor: pointer;
  }
}

:global(html.dark) {
  .alert {
    .x_mark,
    & span {
      color: var(--zinc-200);
    }
  }
}
</style>
