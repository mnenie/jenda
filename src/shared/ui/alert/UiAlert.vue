<script setup lang="ts">
import { computed } from 'vue';
import type { AlertVariant } from './types';
import { variants } from '.';
import { X } from 'lucide-vue-next';
import { useDark } from '@vueuse/core';

const props = withDefaults(
  defineProps<{
    variant?: AlertVariant;
    closable?: boolean;
    content?: string;
  }>(),
  {
    variant: 'default',
    closable: false
  }
);

const emits = defineEmits(['close']);

const isDark = useDark();

const _styles = computed(() => {
  return variants[props.variant];
});

const alertBgColor = computed(() =>
  isDark.value ? _styles.value.dark.background : _styles.value.light.background
);
const alertBorder = computed(() => (isDark.value ? _styles.value.dark.border : _styles.value.light.border));

const valueOfRightPadding = computed(() => props.closable ? '32px' : '10px');
</script>

<template>
  <div :class="$style.alert" :style="{ background: alertBgColor, border: alertBorder }">
    <slot v-if="$slots.default || content">
      <span class="text-sm">{{ content }}</span>
    </slot>
    <X v-if="closable" :class="$style.x_mark" @click="emits('close')" />
  </div>
</template>

<style module lang="scss">
.alert {
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 10px;
  padding: 2px v-bind('valueOfRightPadding') 2px 10px;
  border-radius: 8px;
  margin-bottom: 20px;
  min-height: 22px;

  .x_mark {
    position: absolute;
    top: 50%;
    transform: translateY(-50%);
    right: 10px;
    width: 13px;
    min-width: 13px;
    color: var(--zinc-600);
    cursor: pointer;
  }
}

:global(html.dark) {
  .alert {
    .x_mark,
    & span {
      color: var(--zinc-200);
    }
  }
}
</style>
