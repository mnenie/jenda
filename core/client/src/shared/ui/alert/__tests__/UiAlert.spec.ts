import { shallowMount } from "@vue/test-utils";
import { describe, expect, it } from "vitest";
import { nextTick } from "vue";
import UiAlert from "../UiAlert.vue";

describe("uiAlert", () => {
  const wrapper = shallowMount(UiAlert, {
    props: {
      variant: "default",
    },
  });

  it("should render correctly", () => {
    expect(wrapper.html());
  });

  it("should have content at slot and render x icon", async () => {
    const _w = shallowMount(UiAlert, {
      slots: {
        default: "hello from alert",
      },
      props: {
        closable: true,
      },
    });
    await nextTick();
    expect(_w.html()).toMatchSnapshot();
  });
});
