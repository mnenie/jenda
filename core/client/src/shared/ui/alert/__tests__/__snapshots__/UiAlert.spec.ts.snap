// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`uiAlert > should have content at slot and render x icon 1`] = `
"<div class="alert" style="background: rgb(250, 250, 250); border: 1.5px solid #e4e4e7;">hello from alert<div i-lucide-x="" text-sm="" class="x_mark"></div>
</div>"
`;
